### PR TITLE
Add automatic mode for DAILY_BUILD_ID pass 2

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] If you are adding/removing a .spec file that has multiple-versions supported, please add [@microsoft/cbl-mariner-multi-package-reviewers](https://github.com/orgs/microsoft/teams/cbl-mariner-multi-package-reviewers) team as reviewer [(Eg. golang has 2 versions 1.18, 1.21+)](https://github.com/microsoft/azurelinux/tree/2.0/SPECS/golang)
- [ ] Ready to merge

---

###### Summary 

Refresh/Reimplementation of https://github.com/microsoft/azurelinux/pull/7873.

Based on top of https://github.com/microsoft/azurelinux/pull/9856.

Add a new mechanism to support `DAILY_BUILD_ID=lkg` which automatically updates the toolchain/chroot manifests to match the most recent successful daily build.

The new daily builds will no longer offer the modified `azurelinux-repos` package with modified repos, so we need to use `REPO_LIST=...` to pass the new repo file.

From #7873 I've dropped `DAILY_BUILD_ID=lkg-force`, it just prints an error and tells the user to revert the modified manifests.

**FYI:** I broke up the changes into targeted commits to hopefully make review easier.

###### Change Log  

- Support `DAILY_BUILD_ID=lkg`
- Support using `REPO_LIST` for DAILY_BUILD_ID repo

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

- Local testing
- Testing still in progress

### CODE REVIEW

**Comment by dmcilvaney** [comment]
*NOTE:* Need to take https://github.com/microsoft/azurelinux/pull/9856 first, and re-target this to 3.0-dev.

File: toolkit/Makefile [hunk 1]
```diff
@@ -207,6 +185,13 @@ include $(SCRIPTS_DIR)/build_tag_imagecustomizer.mk
 #   go-tools, clean-go-tools, go-tidy-all (tidy go utilities before committing) go-test-coverage
 include $(SCRIPTS_DIR)/tools.mk
 
+# Based on daily build configuration, configure testing repos
+include $(SCRIPTS_DIR)/preview.mk
+
+# Dynamically update toolchain and chroot manifests for some flows with:
+#   update-toolchain-manifests
```
Comment (dmcilvaney) [inline toolkit/Makefile:192]: No longer used
```suggestion
```

File: toolkit/scripts/daily_build.mk [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+# Contains:
+#	- Configuration for getting daily build packages
+
+# DAILY_BUILD_ID and DAILY_BUILD REPO while similar are mutually exclusive.
+#
+# DAILY_BUILD_ID is points the build tools to the latest daily built
+# packages as part of a tight local workflow.  The DAILY_BUILD_ID argument
+# takes the form: V-v-YYYYMMDD where V-v is the Major-Minor branch
+#
```
Comment (dmcilvaney) [inline toolkit/scripts/daily_build.mk:12]:
```suggestion
# takes the form: V-v-YYYYMMDD where V-v is the Major-Minor branch or
# may be 'lkg' which will auto-detect the latest valid build.
#
```

File: toolkit/scripts/daily_build.mk [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ifneq ($(DAILY_BUILD_ID),)
+    ifeq ($(DAILY_BUILD_ID),lkg)
+        $(call create_folder,$(daily_lkg_workdir))
+
+        override DAILY_BUILD_ID := $(shell $(SCRIPTS_DIR)/get_lkg_id.sh $(daily_lkg_workdir))
+        ifneq ($(.SHELLSTATUS),0)
+            $(error Failed to auto detect DAILY_BUILD_ID)
+        endif
+
+        # We want to be able to auto update the manifest files if we are using lkg
```
Comment (dmcilvaney) [inline toolkit/scripts/daily_build.mk:46]:
```suggestion
```

File: toolkit/scripts/preview.mk [hunk 4]
```diff
@@ -3,44 +3,63 @@
 
 # Contains:
 #	- Create a preview repo file
-#	- Configuration for getting preview packages
+#	- Create a daily build repo file
```
Comment (dmcilvaney) [inline toolkit/scripts/preview.mk:6]:
```suggestion
#     - Configuration for getting preview packages
#	- Create a daily build repo file
```

File: toolkit/docs/building/building.md [hunk 5]
```diff
@@ -479,6 +482,8 @@ If that is not desired all remote sources can be disabled by clearing the follow
 > - `mariner-ms-non-oss.repo` and `mariner-ms-non-oss-preview.repo` - Azure Linux repository containing Microsoft non open Source RPMs with sources not viewable to the public. The preview version serves the same purpose as the official preview repo.
 >
 
+`DAILY_BUILD_ID` and `DAILY_BUILD_REPO` will automiatically augment the `REPO_LIST`.
```
Comment (anphel31) [inline toolkit/docs/building/building.md:485]: typo automiatically

File: toolkit/docs/building/building.md [hunk 6]
```diff
@@ -353,6 +353,8 @@ The `REPO_LIST` variable supports multiple repo files, and they are prioritized
 The Azure Linux base repo is implicitly provided and an optional preview repo is available by setting `USE_PREVIEW_REPO=y`.
 If `DISABLE_UPSTREAM_REPOS=y` is set, any repo that is accessed through the network is disabled.
 
+Daily build packages are avialable via `DAILY_BUILD_ID`. Use `DAILY_BUILD_ID=lkg`, or `DAILY_BUILD_ID=V-v-YYYYMMDD`.
```
Comment (anphel31) [inline toolkit/docs/building/building.md:356]: typo avialable

File: toolkit/scripts/daily_build.mk [hunk 7]
```diff
@@ -0,0 +1,91 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Contains:
+#	- Configuration for getting daily build packages
+
+# DAILY_BUILD_ID and DAILY_BUILD REPO while similar are mutually exclusive.
```
Comment (anphel31) [inline toolkit/scripts/daily_build.mk:7]: typo should be DAILY_BUILD_REPO

File: toolkit/scripts/daily_build.mk [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+# may be 'lkg' which will auto-detect the latest valid build.
+#
+# DAILY_BUILD_REPO also points to the latest daily built packages.  However
+# this is intended for internal teams to access the packages at build time and
+# runtime.  The argument takes the form of a repo file supplied by the
+# Azure Linux Team's daily artifact feed.  The repo file may be provisioned
+# in a downstream image so that it has access to the repo at runtime.  The
+# DAILY_BUILD_REPO argument takes a path to the daily-mariner.repo file.
+#
+# their pipelines on a daily basis. The repo file may be provisioned to their image
```
Comment (anphel31) [inline toolkit/scripts/daily_build.mk:21]: delete this comment? seems out of place

**Review by anphel31 — APPROVED** [review]

File: toolkit/docs/building/building.md (hunk 5 above)
Comment (dmcilvaney) [reply toolkit/docs/building/building.md:485]: Hah, last minute doc updates are a mistake, especially with the spellcheck off.

**Pull request merged**